### DISCUSSION TITLE
Welcome to SHAP Discussions!

### DISCUSSION THREAD

**Discussion by dsgibbons** [dsgibbons/shap#1] [Announcements]
## 👋 Welcome!
SHAP is an essential resource for the ML community, and many projects depend on it. Unfortunately, SHAP's maintainer no longer has the time to keep it up-to-date. This fork attempts to fix SHAP's current issues and merge old PRs. See below for some FAQs. These answers are preliminary and are subject to further discussion:

### What is the current state of SHAP?
The last merged PR was in June 2022, and the issue tracker is inundated with issues that have no replies. 

### How will this be maintained?
For a package as important as SHAP, we want to avoid the mistake of having a single person responsible for the entire repository. Let's create a pool of maintainers so the project is more robust. Two maintainer approvals should be required to merge to master.

### What if slundberg comes back?
I don't know yet. If we make sure there are no breaking changes, perhaps we can get this fork merged with slunberg/shap? If this fork gains enough traction, we may release our own version of shap (name TBD).

### What if slundberg never comes back?
Then we should aim to make our fork the new defacto package for computing Shapley values.

### What needs doing right now?

- [x] Contact issuers of currently open PRs and invite them to contribute to this fork.
- [x] Populate the issue tracker with the most common problems in SHAP.
- [x] Create a pool of maintainers.
- [ ] Create a release strategy.
- [ ] Establish contributing guidelines.
- [x] Add code formatting and linting by default.
- [x] Fix all deprecation warnings.

**Comment by tetelias** [comment]
Looks like slundberg created a PR in his repo couple of days ago so there's some activity however small. Have you contacted him to find out what his long-term plans are?

**Reply by dsgibbons** [reply]
I did. It looks like he is planning to expand the maintainer pool soon. I'll keep this fork open for now in case that doesn't eventuate. I'm hopeful though that he can get some maintainers on board to help him out.

**Reply by Siarkmic** [reply]
I would love to contribute.

**Reply by dsgibbons** [reply]
Feel free to submit a PR or raise an issue @Siarkmic :)

**Reply by connortann** [reply]
Do you have any correspondence from the original author that you are able to share? For example - is it worth explicitly asking what his plans are for SHAP: if he's open to adding maintainers, or if not whether he'd be happy to include a link to this fork on the original repo?

**Reply by dsgibbons** [reply]
I emailed him back in November. He said that the issue tracker had gotten away from him given the demands of his current day-job at Microsoft (which I found interesting since Microsoft seem to use shap for other ventures such as [responsible-ai-toolbox](https://github.com/microsoft/responsible-ai-toolbox), which is actively maintained). He said that he would love to expand maintenance, but when I replied with details he did not respond. Perhaps we can approach him again after we make improvements to this fork?

**Comment by connortann** [comment]
@dsgibbons  what's the latest with SHAP? I haven't seen any activity in almost half a year; is there a larger pool of maintainers now? Or otherwise, shall we start some work on this fork?

**Reply by dsgibbons** [reply]
@connortann no updates from slundberg at this point. It is not clear that the original project will ever be updated again. I'd be happy to start work on this fork, but I only want to start serious work on it if other people are interested in contributing. Would you like to help me start making improvements?

**Reply by connortann** [reply]
I'd definitely be up for helping out, yes! I think there are lots of relatively straightforward changes that we could make to get going.

On **contributors**:  as you say, we need to get a sufficient mass. I think there will be enough people out there, so the challenge will be to advertise this fork adequately and get it to take off. Contacting the authors of PRs on the original repo seems like a good start. Do you think it would be considered rude to merge in changes from other existing forks and PRs?

For **maintainers**: it will be important to nuture a pool of people, rather than a single point of failure. Do you have any thoughts on how to decide who should be a maintainer: perhaps getting people to demonstrate their capability with PRs? An explicit governance doc could help.

**Reply by dsgibbons** [reply]
> On contributors: as you say, we need to get a sufficient mass. I think there will be enough people out there, so the challenge will be to advertise this fork adequately and get it to take off. Contacting the authors of PRs on the original repo seems like a good start. Do you think it would be considered rude to merge in changes from other existing forks and PRs?

I received several responses from previous authors but only had one PR raised. I think, given the amount of time that has passed, we can start merging in those changes. I'd especially like to start by removing deprecation warning and fixing the tests. I hope that, as we make progress, more contributors will be attracted to the project.

> For maintainers: it will be important to nuture a pool of people, rather than a single point of failure. Do you have any thoughts on how to decide who should be a maintainer: perhaps getting people to demonstrate their capability with PRs? An explicit governance doc could help.

Agreed, let's get to work on a CONTRIBUTING.md.

**Reply by thatlittleboy** [reply]
I'm open to help out in (be it as a contributor or as a maintainer) as well, for what it's worth. Considering how many downstream packages rely on the shap package, I think there's real value here in keeping `shap` up-to-date for the entire DS community. If nothing else, at least keep it compatible with latest versions of immediate dependencies (like numpy, matplotlib) and make releases for latest python versions.

I can help to port over some of the existing PRs if the original authors are no longer contactable.

**Reply by dsgibbons** [reply]
@thatlittleboy that sounds great, thank you for your interest! I think it has been long enough that we can start merging the PRs.

**Comment by connortann** [comment]
What governance model would be appropriate for this new fork? There are some good ideas here:
https://opensource.guide/leadership-and-governance/

The original repo probably fits under the "BDFL: Benovelent Dictator for Life" model. Maybe it would make sense for this repo to follow the "Liberal Contribution" model, to support the aim of increasing the pool of maintainers and contributors.

**Reply by dsgibbons** [reply]
Thank you for sharing that link. I like the sound of "Liberal Contribution", and how it couples influence with current contributions instead of historical contributions. That article from Node.js looks like a great start. Perhaps we should create a CONTRIBUTING.md based on the Node.js article? I'm happy for you to raise a PR for this, but I can also write something out over the weekend if you like.

**Comment by dsgibbons** [comment]
This fork was a community-led effort to fix issues in the popular [SHAP repository](https://github.com/slundberg/shap) in 2022/23. At the time, the main repository had stalled. The [project owner](https://github.com/slundberg/) was the sole maintainer, and PRs were blocked by a failing test suite. In June 2023, the main collaborators on this fork (@thatlittleboy, @connortann, and @dsgibbons) were given write-access on the main repository. This fork now just exists as a place for me to work on my own SHAP contributions. Thank you all for your efforts in bringing SHAP back to life!
